catalog/resource_grant.go

```diff
@@ -0,0 +1,179 @@
+	existing, err := a.GetPermissions(securableType, name)
+	if err != nil {
+		return err
+	}
+	err = a.UpdatePermissions(securableType, name, diffPermissionsForPrincipal(principal, list, *existing))
+	if err != nil {
+		return err
+	}
+	return a.WaitForUpdate(1*time.Minute, securableType, name, list, func(current *catalog.PermissionsList, desired catalog.PermissionsList) []catalog.PermissionsChange {
+		return diffPermissionsForPrincipal(principal, desired, *current)
+	})
+}
+
+func filterPermissionsForPrincipal(in catalog.PermissionsList, principal string) (out catalog.PermissionsList) {
+
+	for _, v := range in.PrivilegeAssignments {
+		if v.Principal == principal {
+			out.PrivilegeAssignments = append(out.PrivilegeAssignments, v)
+		}
+	}
+	return
+}
+
+func ResourceGrant() *schema.Resource {
+	s := common.StructToSchema(permissions.UnityCatalogPrivilegeAssignment{},
+		func(m map[string]*schema.Schema) map[string]*schema.Schema {
+
+			m["principal"].ForceNew = true
+
+			allFields := []string{}
+			for field := range permissions.Mappings {
+				allFields = append(allFields, field)
+			}
+			for field := range permissions.Mappings {
+				m[field] = &schema.Schema{
+					Type:          schema.TypeString,
+					Optional:      true,
+					ForceNew:      true,
+					AtLeastOneOf:  allFields,
+					ConflictsWith: permissions.SliceWithoutString(allFields, field),
+				}
+			}
+			return m
+		})
+
+	return common.Resource{
+		Schema: s,
+		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			principal := d.Get("principal").(string)
+			privileges := permissions.ToPrivilegeSlice(d.Get("privileges").(*schema.Set).List())
+			var grants = catalog.PermissionsList{
+				PrivilegeAssignments: []catalog.PrivilegeAssignment{
+					{
+						Principal:  principal,
+						Privileges: privileges,
+					},
+				},
+			}
+			securable, name := permissions.Mappings.KeyValue(d)
+			unityCatalogPermissionsAPI := permissions.NewUnityCatalogPermissionsAPI(ctx, c)
+			err := replacePermissionsForPrincipal(unityCatalogPermissionsAPI, securable, name, principal, grants)
+			if err != nil {
+				return err
+			}
+			d.SetId(fmt.Sprintf("%s/%s", permissions.Mappings.Id(d), principal))
+			return nil
+		},
+		Read: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			principal := d.Get("principal").(string)
+			split := strings.SplitN(d.Id(), "/", 3)
+			if len(split) != 3 {
+				return fmt.Errorf("ID must be three elements split by `/`: %s", d.Id())
+			}
+			grants, err := permissions.NewUnityCatalogPermissionsAPI(ctx, c).GetPermissions(permissions.Mappings.GetSecurableType(split[0]), split[1])
+			grantsForPrincipal := filterPermissionsForPrincipal(*grants, principal)
+			if err != nil {
+				return err
+			}
+			if len(grantsForPrincipal.PrivilegeAssignments) == 0 {
+				return apierr.NotFound("got empty permissions list")
+			}
+			return common.StructToData(grantsForPrincipal, s, d)
```

Comment:
Note: here we're calling StructToData on a different datatype than the one that was used in StructToSchema. This works because the catalog.PermissionsList struct contains a subset of the fields of the UnityCatalogPrivilegeAssignment struct, but this could be a bit confusing. Could we return UnityCatalogPrivilegeAssignment from filterPermissionsForPrincipal instead?

Reply:
Done & added extra defensive check.